exporter/otlpexporter/internal/arrow/stream.go

```diff
@@ -125,18 +138,39 @@ func (s *Stream) run(bgctx context.Context, client arrowpb.ArrowStreamServiceCli
 		s.write(ctx)
 	}()
 
-	if err := s.read(ctx); err != nil {
-		s.telemetry.Logger.Error("arrow recv", zap.Error(err))
-	}
+	// the result from read() is processed after cancel and wait,
+	// so we can set s.client = nil in case of a delayed Unimplemented.
+	err = s.read(ctx)
+
 	// Wait for the writer to ensure that all waiters are known.
 	cancel()
 	ww.Wait()
 
+	if err != nil {
+		// This branch is reached with an unimplemented status
+		// with or without the WaitForReady flag.
+		if status, ok := status.FromError(err); ok && status.Code() == codes.Unimplemented {
+			// This (client == nil) signals the controller
+			// to downgrade when all streams have returned
+			// in that status.
+			//
+			// TODO: Note there are partial failure modes
+			// that will continue to function in a
+			// degraded mode, such as when half of the
+			// streams are successful and half of streams
+			// take this return path.  Design a graceful
+			// recovery mechanism?
+			s.client = nil
+			s.telemetry.Logger.Info("arrow is not supported", zap.Error(err))
+		} else if !errors.Is(err, io.EOF) && !errors.Is(err, context.Canceled) {
```

Comment:
Should we have a final else where we log something about the EOF or cancel?

Reply:
Not sure it will help. The producer of telemetry is going to receive an error in response to their request separately, and this EOF/Cancel is internal to the stream, doesn't directly mean anything is good/bad/incorrect. I could be convinced to add Debug-level statements here (TODOs added), but this makes the test logs look more readable to me.